src/server/game/Movement/MovementGenerators/ChaseMovementGenerator.cpp

```diff
@@ -0,0 +1,185 @@
+    if (!gen || gen->GetMovementGeneratorType() != CHASE_MOTION_TYPE)
+        return false;
+    return (static_cast<ChaseMovementGenerator const*>(gen)->GetTarget() == owner);
+}
+
+static bool PositionOkay(Unit* owner, Unit* target, Optional<float> minDistance, Optional<float> maxDistance, Optional<ChaseAngle> angle)
+{
+    float const distSq = owner->GetExactDistSq(target);
+    if (minDistance && distSq < square(*minDistance))
+        return false;
+    if (maxDistance && distSq > square(*maxDistance))
+        return false;
+    return !angle || angle->IsAngleOkay(target->GetRelativeAngle(owner));
+}
+
+ChaseMovementGenerator::ChaseMovementGenerator(Unit* target, Optional<ChaseRange> range, Optional<ChaseAngle> angle) : AbstractFollower(ASSERT_NOTNULL(target)), _range(range), _angle(angle) {}
+ChaseMovementGenerator::~ChaseMovementGenerator() = default;
+
+void ChaseMovementGenerator::Initialize(Unit* owner)
+{
+    owner->AddUnitState(UNIT_STATE_CHASE);
+    owner->SetWalk(false);
+}
+
+bool ChaseMovementGenerator::Update(Unit* owner, uint32 diff)
+{
+    // owner might be dead or gone (can we even get nullptr here?)
+    if (!owner || !owner->IsAlive())
+        return false;
+
+    // our target might have gone away
+    Unit* const target = GetTarget();
+    if (!target)
+        return false;
+
+    // the owner might've selected a different target (feels like we shouldn't check this here...)
+    if (owner->GetVictim() != target)
+        return false;
+
+    // the owner might be unable to move (rooted or casting), pause movement
+    if (owner->HasUnitState(UNIT_STATE_NOT_MOVE) || owner->IsMovementPreventedByCasting())
+    {
+        owner->StopMoving();
+        return true;
+    }
+
+    bool  const mutualChase     = IsMutualChase(owner, target);
+    float const hitboxSum       = owner->GetCombatReach() + target->GetCombatReach();
+    float const minRange        = _range ? _range->MinRange + hitboxSum : CONTACT_DISTANCE;
+    float const minTarget       = (_range ? _range->MinTolerance : 0.0f) + hitboxSum;
+    float const maxRange        = _range ? _range->MaxRange + hitboxSum : owner->GetMeleeRange(target); // melee range already includes hitboxes
+    float const maxTarget       = _range ? _range->MaxTolerance + hitboxSum : CONTACT_DISTANCE + hitboxSum;
+    Optional<ChaseAngle> angle  = mutualChase ? Optional<ChaseAngle>() : _angle;
+
+    // if we're already moving, periodically check if we're already in the expected range...
+    if (owner->HasUnitState(UNIT_STATE_CHASE_MOVE))
+    {
+        if (_rangeCheckTimer > diff)
+            _rangeCheckTimer -= diff;
+        else
+        {
+            _rangeCheckTimer = RANGE_CHECK_INTERVAL;
+            if (PositionOkay(owner, target, _movingTowards ? Optional<float>() : minTarget, _movingTowards ? maxTarget : Optional<float>(), angle))
+            {
+                owner->StopMoving();
+                _path = nullptr;
```

Comment:
consider maybe adding a return true here?

Reply:
no - if the position is invalid, we want to immediately start pathfinding again (maybe the target moved?)